torch/_prims/utils.py

```diff
@@ -0,0 +1,468 @@
+import torch
+
+from numbers import Number
+from typing import Union, Sequence
+from functools import reduce
+
+
+class TensorMeta(object):
+    """
+    Temporary helper class to model tensor metadata.
+
+    To be replaced with an actual meta tensor.
+    """
+
+    def __init__(
+        self, tensorlike=None, *, shape=None, strides=None, dtype=None, device=None
+    ):
+
+        if isinstance(tensorlike, Number):
+            self.shape = ()
+            self.strides = ()
+            self.dtype = type_to_dtype(type(tensorlike))
+            self.device = torch.device("cpu")
+        elif tensorlike is not None:
+            self.shape = tensorlike.shape
+            self.strides = tensorlike.stride()
+            self.dtype = tensorlike.dtype
+            self.device = tensorlike.device
+        else:
+            assert shape is not None
+            assert strides is not None
+            assert dtype is not None
+            assert device is not None
+
+        self.shape = self.shape if shape is None else shape
+        self.strides = self.strides if strides is None else strides
+        self.dtype = self.dtype if dtype is None else dtype
+        self.device = self.device if device is None else device
+
+        assert isinstance(self.shape, Sequence)
+        assert isinstance(self.strides, Sequence)
+        assert isinstance(self.dtype, torch.dtype)
+        assert isinstance(self.device, torch.device)
+
+        self.ndim = len(self.shape)
+
+    def stride(self):
+        return self.strides
+
+    def numel(self):
+        if len(self.shape) == 0:
+            return 1
+
+        return reduce(lambda x, acc: x * acc, self.shape, 1)
+
+
+TensorLike = (torch.Tensor, TensorMeta)
+
+
+def compare_tensor_meta(a, b):
+    assert isinstance(a, TensorLike)
+    assert isinstance(b, TensorLike)
+
+    for x, y in zip(a.shape, b.shape):
+        assert x == y
+
+    assert a.dtype == b.dtype
+    assert a.device == b.device
+
+
+#
+# Common helper functions
+#
+
+
+def validate_dim_length(length: int):
+    """
+    Validates that an object represents a valid
+    dimension length.
+    """
+
+    assert isinstance(length, int)
+    assert length >= 0
+
+
+def validate_shape(shape: Sequence):
+    """
+    Validates that a sequence represents a valid shape.
+    """
+
+    assert isinstance(shape, Sequence)
+    for l in shape:
+        validate_dim_length(l)
+
+
+def validate_idx(shape: Sequence, idx: int):
+    """
+    Validates that idx is a valid idx for the given shape.
+    """
+
+    assert isinstance(idx, int)
+    assert idx >= 0 and idx < len(shape)
+
+
+def validate_exclusive_idx(shape: Sequence, ex_idx: int):
```

Comment:
Is this new nomenclature? I don't think I've ever heard "exclusive index" before. Exclusive-inclusive would be more accurate given the definition below; another way to think about it is this index is treated as 1-based rather than 0-based (c.f. https://blog.nelhage.com/2015/08/indices-point-between-elements/ )

Reply:
Yeah we can fix the name